File: sydney/Java/JDBC/src/jp/co/ricoh/doquedb/common/Instance.java
// -*-Mode: Java; tab-width: 4; c-basic-offset: 4;-*-
// vi:set ts=4 sw=4:
//
// Instance.java -- ClassIDに相当するクラスのインスタンスを得る
//
// Copyright (c) 2002, 2003, 2004, 2005, 2007, 2023 Ricoh Company, Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package jp.co.ricoh.doquedb.common;

/**
 * {@link ClassID クラスID}に相当するクラスのインスタンスを得る
 *
 */
public final class Instance
{
	/**
	 * {@link ClassID クラスID}に相当するSerializableのサブクラスの
	 * インスタンスを得る
	 *
	 * @param id_	{@link ClassID クラスID}
	 * @return Serializableのサブクラス
	 * @throws java.lang.ClassNotFoundException
	 *			読み込んだ{@link ClassID クラスID}のクラスが見つからない
	 */
	public static Serializable get(int id_)
		throws ClassNotFoundException
	{
		Serializable object = null;
		switch (id_)
		{
		case ClassID.NONE:
			break;
		case ClassID.STATUS:
			object = new Status();
			break;
		case ClassID.INTEGER_DATA:
			object = new IntegerData();
			break;
		case ClassID.UNSIGNED_INTEGER_DATA:
			object = new IntegerData();	//Javaにはunsignedはないので
			break;
		case ClassID.INTEGER64_DATA:
			object = new Integer64Data();
			break;
		case ClassID.UNSIGNED_INTEGER64_DATA:
			object = new Integer64Data(); //Javaにはunsignedはないので
			break;
		case ClassID.FLOAT_DATA:
			object = new FloatData();
			break;
		case ClassID.DOUBLE_DATA:
			object = new DoubleData();
			break;
		case ClassID.DECIMAL_DATA:
			object = new DecimalData();
			break;
		case ClassID.STRING_DATA:
			object = new StringData();
			break;
		case ClassID.DATE_DATA:
			object = new DateData();
			break;
		case ClassID.DATE_TIME_DATA:
			object = new DateTimeData();
			break;
		case ClassID.INTEGER_ARRAY_DATA:
			object = new IntegerArrayData();
			break;
		case ClassID.UNSIGNED_INTEGER_ARRAY_DATA:
			object = new IntegerArrayData();	// Javaにはunsignedはないので
			break;
		case ClassID.STRING_ARRAY_DATA:
			object = new StringArrayData();
			break;
		case ClassID.DATA_ARRAY_DATA:
			object = new DataArrayData();
			break;
		case ClassID.BINARY_DATA:
			object = new BinaryData();
			break;
		case ClassID.NULL_DATA:
			object = NullData.getInstance();
			break;
		case ClassID.EXCEPTION_DATA:
			object = new ExceptionData();
			break;
//		case ClassID.PARAMETER:
//			object = new Parameter();
//			break;
//		case ClassID.BITSET:
//			object = new BitSet();
//			break;
		case ClassID.COMPRESSED_STRING_DATA:
			// CompressedStringDataは伸長されてくるので...
			object = new StringData();
			break;
		case ClassID.COMPRESSED_BINARY_DATA:
			object = new CompressedBinaryData();
			break;
//		case ClassID.OBJECTID_DATA:
//			object = new ObjectIDData();
//			break;
		case ClassID.REQUEST:
			object = new Request();
			break;
		case ClassID.LANGUAGE_DATA:
			object = new LanguageData();
			break;
//		case ClassID.SQL_DATA:
//			object = new SQLData();
//			break;
		case ClassID.COLUMN_META_DATA:
			object = new ColumnMetaData();
			break;
		case ClassID.RESULTSET_META_DATA:
			object = new ResultSetMetaData();
			break;
		case ClassID.WORD_DATA:
			object = new WordData();
			break;
		case ClassID.ERROR_LEVEL:
			object = new ErrorLevel();
			break;
		default:
			throw new ClassNotFoundException();
		}

		return object;
	}
}

//
// Copyright (c) 2002, 2003, 2004, 2005, 2007, 2023 Ricoh Company, Ltd.
// All rights reserved.
//

File: sydney/Python/src/doquedb/common/instance.py
# Copyright (c) 2024 Ricoh Company, Ltd.
#
# Licensed under the Apache License, Version 2.0 (the License);
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
instance.py --  class_idに相当するクラスのインスタンスを得る
"""
from typing import Optional, Any

from ..exception import exceptions
from .constants import ClassID
from .abstracts import Data


class Instance:
    """クラスID(:obj: `ClassID`)に相当するクラスのインスタンスを得る
    """
    # object_map, python_to_trobject_mapをクラス変数で定義する
    object_map = None
    python_to_trobject_map = None

    @staticmethod
    def get(id: int) -> Any:  # 戻り値はSerializableを継承したサブクラス
        """クラスID(:obj: `ClassID`)に相当するSerializableのサブクラスのインスタンスを得る

        Args:
            id (int): クラスID

        Returns:
            Serializable: :obj: `Serializable` のサブクラス

        TODO:
            :dict: ``object_map``をメソッドの外で定義する
        """
        # メソッド外でインポートすると相互参照でエラーになるためメソッド内でインポートする
        from .serialdata import (ColumnMetaData,
                                 ErrorLevel,
                                 ExceptionData,
                                 Request,
                                 Status
                                 )
        from .arraydata import (IntegerArrayData,
                                ResultSetMetaData,
                                StringArrayData,
                                DataArrayData
                                )
        from .scalardata import (FloatData,
                                 NullData,
                                 StringData,
                                 IntegerData,
                                 Integer64Data,
                                 DecimalData,
                                 DoubleData,
                                 DateData,
                                 DateTimeData,
                                 LanguageData,
                                 )
        from .data import BinaryData, WordData

        # object_mapを生成する
        if Instance.object_map is None:
            Instance.object_map = {
                'STATUS': Status,
                'BINARY_DATA': BinaryData,
                'INTEGER_DATA': IntegerData,
                'UNSIGNED_INTEGER_DATA': IntegerData,
                'INTEGER64_DATA': Integer64Data,
                'UNSIGNED_INTEGER64_DATA': Integer64Data,
                'FLOAT_DATA': FloatData,
                'DOUBLE_DATA': DoubleData,
                'DECIMAL_DATA': DecimalData,
                'STRING_DATA': StringData,
                'DATE_DATA': DateData,
                'DATE_TIME_DATA': DateTimeData,
                'INTEGER_ARRAY_DATA': IntegerArrayData,
                'UNSIGNED_INTEGER_ARRAY_DATA': IntegerArrayData,
                'STRING_ARRAY_DATA': StringArrayData,
                'DATA_ARRAY_DATA': DataArrayData,
                'NULL_DATA': NullData,
                'EXCEPTION_DATA': ExceptionData,
                # 'PARAMETER' : Parameter,
                # 'BITSET' : BitSet,
                'COMPRESSED_STRING_DATA': StringData,
                # 'COMPRESSED_BINARY_DATA': CompressedBinaryData,
                # 'OBJECTID_DATA' : ObjectIDData,
                'REQUEST': Request,
                'LANGUAGE_DATA': LanguageData,
                # 'SQL_DATA' : SQLData,
                'COLUMN_META_DATA': ColumnMetaData,
                'RESULTSET_META_DATA': ResultSetMetaData,
                'WORD_DATA': WordData,
                'ERROR_LEVEL': ErrorLevel,
            }

        # ``id``がClassIDに存在するか確認
        object: Optional[Any] = None
        if id == ClassID.NONE.value:
            return object
        elif id in list(map(int, ClassID)):
            # class_idを``id``から逆引き検索
            for item in ClassID:
                if id == item.value:
                    # インスタンス化
                    object = Instance.object_map[item.name]()
        else:
            raise exceptions.UnexpectedError('Class ID is not valid')

        return object

    @staticmethod
    def get_data(type_: type, value: Any) -> Data:
        # メソッド外でインポートすると相互参照でエラーになるためメソッド内でインポートする
        from .scalardata import (NullData,
                                 StringData,
                                 Integer64Data,
                                 DecimalData,
                                 DoubleData,
                                 DateData,
                                 DateTimeData,
                                 LanguageData,
                                 )
        from .data import BinaryData
        from ..driver.dbapi import Date, Timestamp, Language, Decimal, Binary

        # pythonの既定クラスとインスタンスの対応
        # python_to_trobject_mapを生成する
        if Instance.python_to_trobject_map is None:
            Instance.python_to_trobject_map = {
                int: Integer64Data,
                float: DoubleData,
                str: StringData,
                Date: DateData,
                Timestamp: DateTimeData,
                Decimal: DecimalData,
                Language: LanguageData,
                Binary: BinaryData
            }

        if type_ is not type(None):
            if not isinstance(value, type_):
                raise exceptions.UnexpectedError(
                    'expected type is {}, but value has type {}'
                    .format(type_, type(value)))
        else:
            # Noneが与えられた場合はNullDataのインスタンスを返す
            return NullData()

        try:
            object_ = Instance.python_to_trobject_map[type_](value)
        except IndexError as ie:
            raise exceptions.UnexpectedError(f'invalid type: {type_}') from ie

        assert isinstance(object_, Data)
        return object_
